### PR TITLE
 avoid float precision issues leading to overflow in bitmap index 

### PR DESCRIPTION
**Pull request by ngoldbaum**
This fixes the issues @AshKelly ran into with the test SWIFT dataset in #1962.

Ultimately the issue was due to going between float coordinates and int coordinates. The fix is just to avoid those conversions and calculate the int coordinates as we recurse.

In addition we converted some pointer arrays to memoryviews, that makes it easier to debug this code when it breaks.

### CODE REVIEW

**Comment by AshKelly** [comment]
This works perfect on my machine to generate the projection plots which were previously segfaulting. 

The results also agree with the `OffAxisProjection`which was previously, and still is, working.

File: yt/geometry/particle_oct_container.pyx [hunk 1]
```diff
@@ -18,7 +18,7 @@ from oct_container cimport OctreeContainer, Oct, OctInfo, ORDER_MAX, \
     SparseOctreeContainer, OctKey, OctAllocationContainer
 cimport oct_visitors
 from oct_visitors cimport cind, OctVisitor
-from libc.stdlib cimport malloc, free, qsort
+from libc.stdlib cimport malloc, realloc, free, qsort
```
Comment (MatthewTurk) [inline yt/geometry/particle_oct_container.pyx]: I *think* we don't use `realloc`

File: yt/geometry/particle_oct_container.pyx [hunk 2]
```diff
@@ -1846,38 +1806,44 @@ cdef class ParticleBitmapSelector:
...
+                                nlevel, npos, ndds, mi1, ncur_ind)
                     elif nlevel == self.order1:
-                        mi1 = bounded_morton_dds(npos[0], npos[1], npos[2], self.DLE, ndds)
+                        mi1 = encode_morton_64bit(
+                            ncur_ind[0], ncur_ind[1], ncur_ind[2])
                         if sbbox == 2: # an edge cell
                             if self.is_refined(mi1) == 1:
-                                self.recursive_morton_mask(nlevel, npos, ndds, mi1)
+                                self.recursive_morton_mask(
+                                    nlevel, npos, ndds, mi1, zeros)
```
Comment (MatthewTurk) [inline yt/geometry/particle_oct_container.pyx]: I think this is right, but it took me a second to work out that this was not mutated on the other end.

File: yt/geometry/particle_oct_container.pyx (hunk 1 above)
Comment (AshKelly) [reply yt/geometry/particle_oct_container.pyx]: oops, I added this before sending the memoryview refactor to Nathan. I was using it to experiment

File: yt/geometry/particle_oct_container.pyx (hunk 2 above)
Comment (ngoldbaum) [reply yt/geometry/particle_oct_container.pyx]: yup! you did this in a nice clever way without mutating anything :)

Comment (ngoldbaum) [reply yt/geometry/particle_oct_container.pyx]: i guess this is worth a comment though

**Pull request merged**